sycl/plugins/level_zero/pi_level_zero.cpp

```diff
@@ -3505,8 +3505,10 @@ pi_result piKernelCreate(pi_program Program, const char *KernelName,
     return PI_ERROR_UNKNOWN;
   }
 
-  // Update the refcount of the program to show its use by this kernel.
-  PI_CALL(piProgramRetain(Program));
+  // Update the refcount of the program and context to show it's used by this
+  // kernel.
+  { PI_CALL(piProgramRetain(Program)); }
+  { PI_CALL(piContextRetain(Program->Context)); }
```

Comment:
Please explain why this was not needed before, but needed with these changes?

Reply:
In SYCL RT kernels are cached in the context (container in the context class). piContextRelease is called in the ~context() and it is called earlier than cache of the kernels is destructed because of the order of destruction in C++. So, _pi_context object is destroyed ealier than _pi_kernel object.
It was Ok before this change because piKernelRelease didn't use pi_context. Now it uses context to lock the mutex here: https://github.com/intel/llvm/pull/3710/files#diff-15dd1eb076d2164bd9e87d9057f05f652a716498e8cdf5975e564c65309a0985R3757

And that's why with these changes we need to be sure that context is destroyed later than kernel.

Reply:
Great, thanks.